packages/frontend-2/middleware/requireValidWorkspace.ts

```diff
@@ -0,0 +1,56 @@
+import { useApolloClientFromNuxt } from '~~/lib/common/composables/graphql'
+import {
+  convertThrowIntoFetchResult,
+  getFirstErrorMessage
+} from '~~/lib/common/helpers/graphql'
+import { workspaceAccessCheckQuery } from '~~/lib/workspaces/graphql/queries'
+
+/**
+ * Used to validate that the workspace ID refers to a valid workspace and redirects to 404 if not
+ */
+export default defineNuxtRouteMiddleware(async (to) => {
+  const workspaceId = to.params.id as string
+
+  const client = useApolloClientFromNuxt()
+
+  const { data, errors } = await client
+    .query({
+      query: workspaceAccessCheckQuery,
+      variables: { id: workspaceId },
+      context: {
+        skipLoggingErrors: true
+      }
+    })
+    .catch(convertThrowIntoFetchResult)
+
+  if (data?.workspace?.id) return
+
+  const isForbidden = (errors || []).find((e) => e.extensions['code'] === 'FORBIDDEN')
+  const isNotFound = (errors || []).find(
```

Comment:
Have you tested this out? Some of these don't make sense, like there is no error w/ the code `WORKSPACE_NOT_FOUND`.

You should actually try to reach a non-existant workspace or a workspace the user doesn't have access to, and then you'll see the proper error codes you're supposed to use

I checked, the not found error code is: WORKSPACE_NOT_FOUND_ERROR